### PR TITLE
Change `:float` to `{:f, 64}` and add `{:f, 32}` dtype

### PR DESCRIPTION
**Pull request by philss**
The idea is to support more dtypes in the way they are represented in Arrow/Polars.

In the near future, we won't need to "normalise" numeric columns anymore, and by consequence, Explorer is going to use less memory if needed.

This is related to https://github.com/elixir-explorer/explorer/issues/471

### CODE REVIEW

**Comment by philss** [comment]
I feel that we may want to have a "shortcut" version for declaring the dtypes, like in [Nx](https://hexdocs.pm/nx/Nx.html).
We could accept either `:float` for `{:f, 64}` or `:f32` and `:f64`.

Also, the inspection of the series and dataframes could be different, and follow Nx as well - just `f32` or `f64` -, or slightly larger, like `float64`.

WDYT @josevalim  @cigrainger  @billylanchantin ?

**Comment by josevalim** [comment]
I think we should definitely still support `:float` for backwards compatibility. We can add f32/f64 later, it should be a quick addition. Also, please add tests that `to_iovec` with f32 works. Other than that, this is great, ship it!

**Comment by josevalim** [comment]
> I think we should definitely still support :float for backwards compatibility. 

Suggestion: rename `valid_dtype?` to `normalize_dtype` (which returns nil if no dtype) and rename `validate_dtype` to `normalize_dtype`, which returns the normalize dtype. Now change all call sites of `normalize_dtype` and `normalize_dtype!` to consider its return value (which is the normalized dtype). Then it should be easy to add `:float`, but I think we can do that in the next PR. So let's just add the test for `to_iovec` with f32 to this one and the normalize stuff can come in a separate one!

**Comment by cigrainger** [comment]
Yep I agree -- let's keep the 'simple' names like `:float` and give them defaults as they are today. Otherwise this is fantastic. Re: inspection, I prefer to follow Nx as closely as possible.

**Comment by philss** [comment]
@josevalim @cigrainger I made changes applying your suggestions. Please take a look.

The only thing I didn't change was to add the aliases `:f32` and `:f64` for the float dtypes. I'm planning to do that in another PR tomorrow, with docs as well.

**Review by josevalim — APPROVED** [review]

File: README.md [hunk 1]
```diff
@@ -13,7 +13,7 @@ data exploration to Elixir.
 
 Explorer high-level features are:
 
-- Simply typed series: `:binary`, `:boolean`, `:category`, `:date`, `:datetime`, `:duration`, `:float`, `:integer`, `:string`, and `:time`.
+- Simply typed series: `:binary`, `:boolean`, `:category`, `:date`, `:datetime`, `:duration`, `{:f, 32}`, `{:f, 64}`, `:integer`, `:string`, and `:time`.
```
Comment (cigrainger) [inline README.md:16]: I wonder if we shouldn't do something like:
```suggestion
- Simply typed series: `:binary`, `:boolean`, `:category`, `:date`, `:datetime`, `:duration`, `:float` (`{:f, 32}` or `{:f, 64}`), `:integer`, `:string`, and `:time`.
```

WDYT?

**Review by cigrainger — APPROVED** [review]
Dropped one Q but otherwise this looks fantastic. I'm excited to support more explicit datatypes, esp w/ zero-copy conversion to tensors.

File: README.md (hunk 1 above)
Comment (philss) [reply README.md]: Agreed! I added more mentions of it in other places as well: https://github.com/elixir-explorer/explorer/pull/739/commits/3605cfcc7ca6877ea7568626d91b933e97148068

**Pull request merged**